Code/GraphMol/MolStandardize/Fragment.cpp

```diff
@@ -179,57 +179,84 @@ LargestFragmentChooser::LargestFragmentChooser(
   countHeavyAtomsOnly = other.countHeavyAtomsOnly;
 }
 
-ROMol *LargestFragmentChooser::choose(const ROMol &mol) {
+ROMol *LargestFragmentChooser::choose(const ROMol &mol) const {
+  auto res = new RWMol(mol);
+  chooseInPlace(*res);
+  // resanitize the molecule
+  MolOps::sanitizeMol(*res);
+
+  return static_cast<ROMol *>(res);
+}
+
+void LargestFragmentChooser::chooseInPlace(RWMol &mol) const {
   BOOST_LOG(rdInfoLog) << "Running LargestFragmentChooser\n";
 
   if (!mol.getNumAtoms()) {
-    return new ROMol(mol);
+    return;
+  }
+
+  std::vector<std::vector<int>> frags;
+  MolOps::getMolFrags(mol, frags);
+  if (frags.size() == 1) {
+    // nothing to do
+    return;
   }
-  std::vector<boost::shared_ptr<ROMol>> frags = MolOps::getMolFrags(mol);
+
   LargestFragmentChooser::Largest l;
 
-  for (const auto &frag : frags) {
-    std::string smiles = MolToSmiles(*frag);
+  SmilesWriteParams ps;
+  int bestFragment = -1;
+  for (auto fidx = 0u; fidx < frags.size(); ++fidx) {
+    const auto &frag = frags[fidx];
+    std::string smiles = MolFragmentToSmiles(mol, ps, frag);
     BOOST_LOG(rdInfoLog) << "Fragment: " << smiles << "\n";
-    bool organic = isOrganic(*frag);
+    bool organic = isOrganic(mol, frag);
     if (this->preferOrganic) {
       // Skip this fragment if not organic and we already have an organic
       // fragment as the largest so far
-      if (l.Fragment != nullptr && l.Organic && !organic) {
+      if (bestFragment >= 0 && l.Organic && !organic) {
         continue;
       }
       // Reset largest if it wasn't organic and this fragment is organic
       // if largest and organic and not largest['organic']:
-      if (l.Fragment != nullptr && organic && !l.Organic) {
-        l.Fragment = nullptr;
+      if (bestFragment >= 0 && organic && !l.Organic) {
+        bestFragment = -1;
       }
     }
     unsigned int numatoms = 0;
     if (this->useAtomCount) {
-      for (const auto at : frag->atoms()) {
+      for (const auto idx : frag) {
         ++numatoms;
         if (!this->countHeavyAtomsOnly) {
-          numatoms += at->getTotalNumHs();
+          numatoms += mol.getAtomWithIdx(idx)->getTotalNumHs();
         }
       }
       // Skip this fragment if fewer atoms than the largest
-      if (l.Fragment != nullptr && (numatoms < l.NumAtoms)) {
+      if (bestFragment >= 0 && (numatoms < l.NumAtoms)) {
         continue;
       }
     }
 
     // Skip this fragment if equal number of atoms but weight is lower
-    double weight = Descriptors::calcExactMW(*frag);
-    if (l.Fragment != nullptr &&
-        (!this->useAtomCount || numatoms == l.NumAtoms) &&
+    double weight = 0;
+    for (auto idx : frag) {
+      const auto atom = mol.getAtomWithIdx(idx);
+      // it's not important to be perfect here
+      weight += 100 * atom->getAtomicNum() + atom->getIsotope() -
+                atom->getFormalCharge() * .1;
```

Comment:
Making weight depend on formal charge looks a bit of a hack, and could lead to unintentional ties (10 formal charges count like one neutron). Wouldn't it be better to leave formal charge out of the weight computation and rather add it as an additional compare criterion?

Reply:
this is clearly a heuristic (sounds nicer than "hack") and it doesn't have to be exactly correct, just reproducible.

Plus, how often are you going to have a molecule with two fragments which have the same number of atoms and the same atoms but which differ by one neutron and a net formal charge difference of 10?

I think it's a theoretical risk. :-)

Reply:
It is a hack :-) but I am fine with it.